File: src/main/python/AoC2024_09.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 9
#

import heapq
import sys
from enum import Enum
from enum import auto
from enum import unique

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Input = list[int]
Output1 = int
Output2 = int
File = tuple[int, int, int]

TRIANGLE = [0, 0, 1, 3, 6, 10, 15, 21, 28, 36]

TEST = """\
2333133121414131402
"""


@unique
class Mode(Enum):
    MODE_1 = auto()
    MODE_2 = auto()

    def create_files(self, fid: int, pos: int, sz: int) -> list[File]:
        match self:
            case Mode.MODE_1:
                return [(fid, pos + i, 1) for i in range(sz)]
            case Mode.MODE_2:
                return [(fid, pos, sz)]

    def checksum(self, f: File) -> int:
        fid, pos, sz = f
        match self:
            case Mode.MODE_1:
                return fid * pos
            case Mode.MODE_2:
                return fid * (pos * sz + TRIANGLE[sz])


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return list(map(int, next(iter(input_data))))

    def solve(self, disk: Input, mode: Mode) -> int:
        files = list[File]()
        free_by_sz: list[list[int]] = [[] for _ in range(10)]
        is_free, fid, pos = False, 0, 0
        for n in disk:
            if is_free:
                heapq.heappush(free_by_sz[n], pos)
            else:
                files.extend(mode.create_files(fid, pos, n))
                fid += 1
            pos += n
            is_free = not is_free
        ans = 0
        for fid, pos, sz in reversed(files):
            earliest = min(
                (
                    (i, free)
                    for i, free in enumerate(free_by_sz[sz:], sz)
                    if len(free) > 0
                ),
                key=lambda e: e[1][0],
                default=None,
            )
            new_pos = pos
            if earliest is not None:
                free_sz, free = earliest
                free_pos = free[0]
                if free_pos < pos:
                    heapq.heappop(free_by_sz[free_sz])
                    new_pos = free_pos
                    if sz < free_sz:
                        heapq.heappush(free_by_sz[free_sz - sz], new_pos + sz)
            ans += mode.checksum((fid, new_pos, sz))
        return ans

    def part_1(self, disk: Input) -> Output1:
        return self.solve(disk, Mode.MODE_1)

    def part_2(self, disk: Input) -> Output2:
        return self.solve(disk, Mode.MODE_2)

    @aoc_samples(
        (
            ("part_1", TEST, 1928),
            ("part_2", TEST, 2858),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 9)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_09.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;
import static com.github.pareronia.aoc.IterTools.enumerateFrom;

import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

import com.github.pareronia.aoc.Utils;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2024_09
        extends SolutionBase<int[], Long, Long> {

    public static final long[] TRIANGLE = {0, 0, 1, 3, 6, 10, 15, 21, 28, 36};
    
    private AoC2024_09(final boolean debug) {
        super(debug);
    }
    
    public static AoC2024_09 create() {
        return new AoC2024_09(false);
    }
    
    public static AoC2024_09 createDebug() {
        return new AoC2024_09(true);
    }
    
    @Override
    protected int[] parseInput(final List<String> inputs) {
        return Utils.asCharacterStream(inputs.get(0))
                .mapToInt(ch -> ch - '0')
                .toArray();
    }
    
    public long solve(final int[] input, final Mode mode) {
        final List<File> files = new ArrayList<>(input.length * 10);
        final List<PriorityQueue<Integer>> freeBySize = new ArrayList<>(10);
        for (int i = 0; i < 10; i++) {
            freeBySize.add(new PriorityQueue<>());
        }
        boolean isFree = false;
        int id = 0;
        int pos = 0;
        for (final int n : input) {
            if (isFree) {
                freeBySize.get(n).add(pos);
            } else {
                files.addAll(mode.createFiles(id, pos, n));
                id++;
            }
            pos += n;
            isFree = !isFree;
        }
        long ans = 0;
        for (int j = files.size() - 1; j >= 0; j--) {
            final File file = files.get(j);
            enumerateFrom(file.size, freeBySize.stream().skip(file.size))
                .filter(e -> !e.value().isEmpty())
                .min((e1, e2) -> Integer.compare(e1.value().peek(), e2.value().peek()))
                .filter(e -> e.value().peek() < file.pos)
                .ifPresent(e -> {
                    final int free_size = e.index();
                    final int free_pos = e.value().poll();
                    file.pos = free_pos;
                    if (file.size < free_size) {
                        freeBySize.get(free_size - file.size)
                            .add(file.pos + file.size);
                    }
                });
            ans += mode.fileChecksum(file);
        }
        return ans;
    }
    
    @Override
    public Long solvePart1(final int[] input) {
        return solve(input, Mode.MODE_1);
    }

    @Override
    public Long solvePart2(final int[] input) {
        return solve(input, Mode.MODE_2);
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "1928"),
        @Sample(method = "part2", input = TEST, expected = "2858"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2024_09.create().run();
    }

    private static final String TEST = "2333133121414131402";

    private static class File {
        int id;
        int pos;
        int size;
        
        public File(final int id, final int pos, final int size) {
            this.id = id;
            this.pos = pos;
            this.size = size;
        }
    }

    private enum Mode {
        MODE_1,
        MODE_2;

        public List<File> createFiles(
                final int id, final int pos, final int size
        ) {
            return switch (this) {
                case MODE_1 -> range(size).intStream()
                                    .mapToObj(i -> new File(id, pos + i, 1))
                                    .toList();
                case MODE_2 -> List.of(new File(id, pos, size));
            };
        }

        public long fileChecksum(final File f) {
            return switch (this) {
                case MODE_1 -> (long) f.id * f.pos;
                case MODE_2 -> f.id * (f.pos * f.size + TRIANGLE[f.size]);
            };
        }
    }
}